cc_core.c

```diff
@@ -992,12 +992,173 @@ void general_recursion(FUNCTION f, char* s, char* name, FUNCTION iterate)
+
+		emit_out("!0 R1 LOAD32 R15 MEMORY\n~0 JUMP_ALWAYS\n%");
+		emit_out(int2str(current_target->type->size, 10, FALSE));
+		emit_out("\n");
+
+		emit_out("'9' R0 '0' R1 MUL R0 ARITH2_ALWAYS\n");
+		emit_out("{R1} POP_ALWAYS\n");
+	}
+	else if(AARCH64 == Architecture)
+	{
+		emit_out("PUSH_X1\n");
+
+		emit_out("LOAD_W0_AHEAD\nSKIP_32_DATA\n%");
+		emit_out(int2str(current_target->type->size, 10, FALSE));
+		emit_out("\n");
+
+		emit_out("MUL_X0_X1_X0\n");
+		emit_out("POP_X1\n");
+	}
+	else if((RISCV32 == Architecture) || (RISCV64 == Architecture))
+	{
+		if(RISCV32 == Architecture) emit_out("rs1_sp rs2_a1 @-4 sw\n");
+		else if(RISCV64 == Architecture) emit_out("rs1_sp rs2_a1 @-8 sd\n");
+
+		emit_out("rd_a0 !");
+		emit_out(int2str(current_target->type->size, 10, FALSE));
+		emit_out(" addi\n");
+
+		emit_out("rd_a0 rs1_a1 rs2_a0 mul\n");
+
+		if(RISCV32 == Architecture) emit_out("rd_a1 rs1_sp !-4 lw\n");
+		else if(RISCV64 == Architecture) emit_out("rd_a1 rs1_sp !-8 ld\n");
+	}
+
+	emit_out("# pointer arithmetic end\n");
+}
+
+int operand_is_pointer(struct token_list* operand)
+{
+	if(operand->s[0] == ')')
+	{
+		while (operand->s[0] != '(')
+		{
+			operand = operand->prev;
+		}
+		operand = operand->prev; /* get the function name */
+
+		if(match(operand->s, "sizeof"))
+		{
+			return FALSE;
+		}
+
+		if(in_set(operand->s[0], "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_"))
+		{
+			/* Confusingly the global_function_list does not contain the return type
+			 * of functions so we'll need the global_symbol_list */
+			struct token_list* a = sym_lookup(operand->s, global_symbol_list);
+			require(NULL != a, operand->s);
+
+			return a->type->type != a->type;
+		}
+	}
+	else if (operand->s[0] == ']')
+	{
+		return FALSE;
+	}
+
+	struct token_list* a = sym_lookup(operand->s, global_constant_list);
+	if(NULL != a) return FALSE;
+
+	a = sym_lookup(operand->s, function->locals);
+	if(NULL != a)
+	{
+		require(NULL != a->type, "function locals type NULL");
+		return a->type->type != a->type;
+	}
+
+	a = sym_lookup(operand->s, function->arguments);
+	if(NULL != a){
+		require(NULL != a->type, "function arguments type NULL");
+		return a->type->type != a->type;
+	}
+
+	a = sym_lookup(operand->s, global_symbol_list);
+	if(NULL != a){
+		require(NULL != a->type, "global symbol list type NULL");
+		return a->type->type != a->type;
+	}
+
+	return !in_set(operand->s[0], "0123456789");
+}
+
 void arithmetic_recursion(FUNCTION f, char* s1, char* s2, char* name, FUNCTION iterate)
 {
 	require(NULL != global_token, "Received EOF in arithmetic_recursion\n");
 	if(match(name, global_token->s))
 	{
 		common_recursion(f);
+		if(name[0] == '+' || name[0] == '-')
+		{
+			if(current_target->type != current_target)
+			{
+				struct token_list* second_operand = global_token->prev;
+				struct token_list* operator = second_operand->prev;
+
+				while (operator->s[0] != '+' && !match(operator->s, "-"))
+				{
+					operator = operator->prev;
+				}
```

Comment:
What's the purpose of this bit?

Reply:
In the case of `a[0] + a[1]` `second_operand` would be the second `]` and we would need to backtrack until we find the operator (`+`) and we can then go back one more in order to get the last token of the first operand`.

Reply:
Gotcha!